File: app/lib/domain/day_history.dart
/// Screens 47/48 — «История дня».
///
/// The day's trail as the app reads it: where she went, how far, and what
/// happened. The server does the thinning and the arithmetic (see
/// packages/backend/src/safety/dayHistory.ts) so that the number printed here
/// is the length of the line the map draws — this file parses and formats, and
/// deliberately does not compute a second opinion about either.
library;

import 'dart:convert';

/// One point on the drawn line.
class RoutePoint {
  final DateTime at;
  final double lat;
  final double lng;

  const RoutePoint({required this.at, required this.lat, required this.lng});

  static RoutePoint? fromJson(Map<String, dynamic> j) {
    final at = DateTime.tryParse('${j['observedAt']}');
    final coords = j['coords'];
    if (at == null || coords is! Map) return null;
    final lat = (coords['lat'] as num?)?.toDouble();
    final lng = (coords['lng'] as num?)?.toDouble();
    if (lat == null || lng == null) return null;
    return RoutePoint(at: at.toLocal(), lat: lat, lng: lng);
  }
}

enum DayEventKind { enter, exit, sos }

/// Something that happened during the day.
class DayEvent {
  final DateTime at;
  final DayEventKind kind;

  /// The zone, for a crossing. Null for an SOS — and null for a crossing whose
  /// zone has since been deleted, which the screen has to be able to show
  /// rather than pretend the crossing did not happen.
  final String? zoneName;
  final double? lat;
  final double? lng;

  /// What this SOS was marked as, if it has been closed. Null on every crossing
  /// and on an SOS nobody has answered for yet — which the detail screen shows
  /// as no chip selected, rather than guessing one.
  final SosOutcome? outcome;

  const DayEvent({
    required this.at,
    required this.kind,
    this.zoneName,
    this.lat,
    this.lng,
    this.outcome,
  });

  /// The same event with a different verdict, for updating the list in place
  /// after a save without refetching the day.
  DayEvent withOutcome(SosOutcome o) => DayEvent(
        at: at, kind: kind, zoneName: zoneName, lat: lat, lng: lng, outcome: o,
      );

  static DayEvent? fromJson(Map<String, dynamic> j) {
    final at = DateTime.tryParse('${j['at']}');
    if (at == null) return null;
    final kind = switch ('${j['kind']}') {
      'enter' => DayEventKind.enter,
      'exit' => DayEventKind.exit,
      'sos' => DayEventKind.sos,
      // An unknown kind is a server newer than this build. Dropping the event
      // is better than crashing the screen, and better than inventing a kind.
      _ => null,
    };
    if (kind == null) return null;
    final zone = j['zoneName'];
    return DayEvent(
      at: at.toLocal(),
      kind: kind,
      zoneName: zone is String && zone.isNotEmpty ? zone : null,
      lat: (j['lat'] as num?)?.toDouble(),
      lng: (j['lng'] as num?)?.toDouble(),
      outcome: sosOutcomeFromWire(j['outcome']),
    );
  }
}

/// One day of a child's movements.
class DayHistory {
  /// The calendar day this describes, as YYYY-MM-DD.
  final String day;
  final List<RoutePoint> points;
  final int distanceM;

  /// Fixes recorded before the trail was simplified for drawing.
  final int rawCount;
  final List<DayEvent> events;

  /// How long ROUTE POINTS are kept on the server — read from the constant the
  /// sweep uses rather than typed here, so the screen cannot promise a window
  /// the sweep does not keep.
  ///
  /// It covers `location_history` and nothing else. The crossings and the SOS
  /// rows in [events] come from other tables that no sweep touches, and the
  /// screen says so separately; do not let this number spread over them.
  final int retentionDays;

  const DayHistory({
    required this.day,
    required this.points,
    required this.distanceM,
    required this.rawCount,
    required this.events,
    required this.retentionDays,
  });

  /// Nothing was recorded. Distinct from a failed load, which the screen shows
  /// differently — a tracker that was switched off is not an error.
  bool get isEmpty => points.isEmpty && events.isEmpty;

  /// The «4 точки» of the badge — how many are drawn.
  int get pointCount => points.length;

  /// The line on screen is fewer points than the tracker reported. True when
  /// that is worth saying, so the badge's small number cannot be mistaken for
  /// a tracker that barely reported.
  bool get wasSimplified => rawCount > points.length;

  static DayHistory fromJson(Map<String, dynamic> j) => DayHistory(
        day: '${j['day']}',
        points: [
          for (final p in (j['points'] as List? ?? const []))
            if (p is Map<String, dynamic>)
              if (RoutePoint.fromJson(p) case final r?) r,
        ],
        distanceM: (j['distanceM'] as num?)?.round() ?? 0,
        rawCount: (j['rawCount'] as num?)?.toInt() ?? 0,
        events: [
          for (final e in (j['events'] as List? ?? const []))
            if (e is Map<String, dynamic>)
              if (DayEvent.fromJson(e) case final d?) d,
        ],
        retentionDays: (j['retentionDays'] as num?)?.toInt() ?? 90,
      );

  static DayHistory parse(String body) =>
      fromJson(jsonDecode(body) as Map<String, dynamic>);
}

/// «3,2 км» / «450 м».
///
/// A comma, because that is the decimal separator in both Russian and Kazakh,
/// and one decimal place: «3,17 км» claims a precision GPS does not have over a
/// day's walk.
String formatDistance(int metres) {
  if (metres < 1000) return '$metres м';
  final km = metres / 1000;
  return '${km.toStringAsFixed(1).replaceAll('.', ',')} км';
}

/// What a parent can mark an SOS as afterwards — the four chips of screen 48.
///
/// Mirrors SOS_OUTCOMES on the server. Four, and no free-text box: a parent
/// closing an alarm at midnight will not write a paragraph, and an outcome
/// field that is usually empty tells nobody anything.
enum SosOutcome { falsePress, scared, neededHelp, unknown }

/// Parse the server's key. Null for absent, and null for a key this build does
/// not know — a newer server is not a reason to crash the day, and inventing a
/// verdict is worse than showing none.
SosOutcome? sosOutcomeFromWire(Object? v) => switch (v) {
      'false_press' => SosOutcome.falsePress,
      'scared' => SosOutcome.scared,
      'needed_help' => SosOutcome.neededHelp,
      'unknown' => SosOutcome.unknown,
      _ => null,
    };

extension SosOutcomeWire on SosOutcome {
  String get wire => switch (this) {
        SosOutcome.falsePress => 'false_press',
        SosOutcome.scared => 'scared',
        SosOutcome.neededHelp => 'needed_help',
        SosOutcome.unknown => 'unknown',
      };

  /// The l10n key for the chip's label.
  String get l10nKey => switch (this) {
        SosOutcome.falsePress => 'sos_out_false',
        SosOutcome.scared => 'sos_out_scared',
        SosOutcome.neededHelp => 'sos_out_help',
        SosOutcome.unknown => 'sos_out_unknown',
      };
}

File: packages/backend/src/safety/dayHistory.ts
/**
 * Frames 47/48 — «История дня» and «Событие из истории дня».
 *
 * docs/CLAUDE-app-design.md: «карта маршрута с точками + плашка «3,2 км ·
 * 4 точки» → таймлайн событий (вышла, пришла в школу, SOS, дома) → плашка
 * «маршруты хранятся 90 дней».»
 *
 * The trail has been written on every fix since the first one and pruned at
 * ninety days, and between the two there was no way to look at it. This turns
 * the raw fixes into the two things the screen shows: a line worth drawing, and
 * a list of what happened.
 *
 * ON SIMPLIFYING. A tracker reporting every thirty seconds makes 2 880 points a
 * day. Drawn whole that is a smear, and sent whole it is half a megabyte over a
 * phone connection. The trail is thinned by distance rather than by taking
 * every Nth point: dropping every other fix loses the corner where she turned,
 * which is the part of a route anybody actually reads.
 *
 * ON DISTANCE. Summed over the THINNED points, so the figure on the screen is
 * the length of the line the screen draws. Summing the raw fixes gives a larger
 * number — GPS jitter while sitting still adds up to hundreds of metres an hour
 * — and «3,2 км» beside a line that is plainly one kilometre reads as a bug in
 * whichever number the reader trusts less.
 *
 * PURE: fixes and events in, a day's story out.
 */

export interface Fix {
  observedAt: string;
  coords: { lat: number; lng: number; accuracyM?: number };
}

/** A geofence crossing, as the events table records it. */
export interface Crossing {
  at: string;
  transition: 'enter' | 'exit';
  /** The zone's name, for «пришла в школу». Null if the zone is gone. */
  zoneName: string | null;
}

/** An SOS press. */
export interface SosPress {
  at: string;
  lat?: number;
  lng?: number;
  /** What the parent marked it as afterwards; null while still open. */
  outcome?: SosOutcome | null;
}

export type DayEventKind = 'enter' | 'exit' | 'sos';

export interface DayEvent {
  at: string;
  kind: DayEventKind;
  /** Zone name for a crossing; null for an SOS, which happens anywhere. */
  zoneName: string | null;
  lat?: number;
  lng?: number;
  /**
   * SOS only, and only once closed. The detail screen preselects the chip from
   * this, so reopening an alarm shows the verdict already recorded rather than
   * asking again — and asking again is how a second answer overwrites a first.
   */
  outcome?: SosOutcome | null;
}

export interface DayHistory {
  /** The line to draw, oldest first. */
  points: Fix[];
  /** Metres along the drawn line. */
  distanceM: number;
  /** How many points are drawn — the «4 точки» in the badge. */
  pointCount: number;
  /** What happened, oldest first. */
  events: DayEvent[];
  /** Fixes before thinning, so the screen can be honest about simplifying. */
  rawCount: number;
}

const EARTH_R = 6_371_000;

/** Metres between two coordinates. Haversine — good to a metre at city scale. */
export function metresBetween(
  a: { lat: number; lng: number },
  b: { lat: number; lng: number },
): number {
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const la1 = toRad(a.lat);
  const la2 = toRad(b.lat);
  const h = Math.sin(dLat / 2) ** 2 + Math.cos(la1) * Math.cos(la2) * Math.sin(dLng / 2) ** 2;
  return 2 * EARTH_R * Math.asin(Math.min(1, Math.sqrt(h)));
}

/**
 * Default thinning distance. Below this a phone is standing still as far as a
 * city map is concerned, and the fixes are jitter.
 */
export const MIN_STEP_M = 40;

/**
 * Keep the first fix, the last, and every one at least [minStepM] from the last
 * one kept.
 *
 * Distance-based rather than every-Nth on purpose: taking every tenth point
 * drops the corner where she turned and keeps ten identical points from the
 * half hour she sat in class. The LAST point is always kept — it is where she
 * is now, and it is the one thing on this screen somebody is looking for.
 */
export function thinTrail(fixes: Fix[], minStepM = MIN_STEP_M): Fix[] {
  if (fixes.length <= 2) return [...fixes];
  const out: Fix[] = [fixes[0]];
  for (let i = 1; i < fixes.length - 1; i++) {
    if (metresBetween(out[out.length - 1].coords, fixes[i].coords) >= minStepM) {
      out.push(fixes[i]);
    }
  }
  out.push(fixes[fixes.length - 1]);
  return out;
}

export function buildDayHistory(input: {
  fixes: Fix[];
  crossings: Crossing[];
  sos: SosPress[];
  minStepM?: number;
}): DayHistory {
  // Sorted here rather than trusted: fixes arrive out of order whenever an
  // offline tracker flushes its buffer, and a trail drawn in arrival order is
  // a line that doubles back on itself.
  const fixes = [...input.fixes].sort((a, b) => a.observedAt.localeCompare(b.observedAt));
  const points = thinTrail(fixes, input.minStepM ?? MIN_STEP_M);

  let distanceM = 0;
  for (let i = 1; i < points.length; i++) {
    distanceM += metresBetween(points[i - 1].coords, points[i].coords);
  }

  const events: DayEvent[] = [
    ...input.crossings.map((c) => ({
      at: c.at,
      kind: c.transition as DayEventKind,
      zoneName: c.zoneName,
    })),
    ...input.sos.map((s) => ({
      at: s.at,
      kind: 'sos' as const,
      zoneName: null,
      lat: s.lat,
      lng: s.lng,
      outcome: s.outcome ?? null,
    })),
  ].sort((a, b) => a.at.localeCompare(b.at));

  return {
    points,
    distanceM: Math.round(distanceM),
    pointCount: points.length,
    events,
    rawCount: fixes.length,
  };
}

/**
 * Frame 48 — «Чем закончилось». What a parent can mark an SOS as, afterwards.
 *
 * Four, and no free-text box. A parent closing an alarm at midnight will not
 * write a paragraph, and an outcome field that is usually empty tells nobody
 * anything. These four are the ones that change what we would do next time.
 */
export const SOS_OUTCOMES = [
  { key: 'false_press', ru: 'Случайное нажатие', kk: 'Кездейсоқ басу' },
  { key: 'scared', ru: 'Испугалась, всё хорошо', kk: 'Қорықты, бәрі жақсы' },
  { key: 'needed_help', ru: 'Нужна была помощь', kk: 'Көмек қажет болды' },
  { key: 'unknown', ru: 'Не удалось выяснить', kk: 'Анықтай алмадық' },
] as const;

export type SosOutcome = (typeof SOS_OUTCOMES)[number]['key'];

export function isSosOutcome(v: unknown): v is SosOutcome {
  return typeof v === 'string' && SOS_OUTCOMES.some((o) => o.key === v);
}
